---
title: "R_assignment"
output: html_document
---

## R Markdown
This is yawei's R assignment review
My comments are written next to ###LF throughout the scripts.
My R assignment will consist of three parts:
Inspect and analysis two data files. 
Additional analysis and visualization.
Reviewing two assignments from my two peers in class.
#loading fang_et_al_genotypes&snp_position
```{r}
library(tidyverse)
library(RCurl)
x <- getURL('https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2019/master/assignments/UNIX_Assignment/fang_et_al_genotypes.txt')
fang_et_al_genotypes <- read.delim(text = x)
y <- getURL('https://raw.githubusercontent.com/EEOB-BioData/BCB546X-Fall2019/master/assignments/UNIX_Assignment/snp_position.txt')
snp_position <- read.delim(text = y)


###LF: I can't read the files because you put the absolute path, so I changed this so that I can read the files
```
#Data inspection
#file.size
#the structure of the dataset
#number of columns, number of rows
```{r}
file.size("fang_et_al_genotypes")
file.size("snp_position")
dim(fang_et_al_genotypes)
dim(snp_position)
colnames(fang_et_al_genotypes)
colnames(snp_position)
rownames(fang_et_al_genotypes)
rownames(snp_position)
```
#Data processing

#for the fang

#filter the group into both maize and teosinte
```{r}
maize_fang_et_al_genotypes <- filter(fang_et_al_genotypes, Group == "ZMMLR" | Group == "ZMMIL" | Group == "ZMMMR")

teosinte_fang_et_al_genotypes <- filter(fang_et_al_genotypes, Group == "ZMPBA" | Group == "ZMPIL" | Group == "ZMPJA")
```


#we need to remove the first three columnes of the two groups.
```{r}
maize_fang_et_al_genotypes <- maize_fang_et_al_genotypes[,-c(1,2,3)]

teosinte_fang_et_al_genotypes <- teosinte_fang_et_al_genotypes[,-c(1,2,3)]
```

#Transpose the dataset.
```{r}
maize_fang_et_al_genotypes <- t (maize_fang_et_al_genotypes)
teosinte_fang_et_al_genotypes <- t (teosinte_fang_et_al_genotypes)
```

#for the snp
#we need to extract the columnes of (1,3,4) from the data of snp
```{r}
snp_position <- snp_position[,-c(2,5:15)]
```


#remove the first row of SNP 


#Then we can combine both fang and snp files.
```{r}

maize_fang_et_al_genotypes <- cbind(snp_position, maize_fang_et_al_genotypes)

teosinte_fang_et_al_genotypes <- cbind(snp_position, teosinte_fang_et_al_genotypes)

###LF: You can also use merge()
```



#Now we can start to creat files we want.
#For maize (Group = ZMMIL, ZMMLR, and ZMMMR in the third column of the fang_et_al_genotypes.txt file) we want 20 files in total:

#10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?
```{r}
for (i in 1:10) {
  maize <- filter(maize_fang_et_al_genotypes, Chromosome == i)
  maize <- rename( maize,NA,"?")
  maize <- arrange(maize, Position)
  outpath <- "/Users/yaweili/yawei-R-Assignment/maize_increasing_data/"
  nam <- sapply(
    names(maize),function(x){
                    paste("maize_in", i, ".csv", sep='')
    })
  out_filePath <- sapply(nam, function(x){
                     paste(outpath, x, sep='/')})
  write.csv(maize, file=out_filePath[i])
}

###LF: I'm not sure if it's very effective, but I recommend using different names for the new data frames to be able to see/fix errors
```


#10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -


#1 file with all SNPs with unknown positions in the genome (these need not be ordered in any particular way)
```{r}
for (i in 1:10) {
  maize <- filter(maize_ded, Chromosome == i)
  maize <- arrange(maize, desc(Position))
  outpath <- "/Users/yaweili/yawei-R-Assignment/maize_decreasing_data/"
  nam <- sapply(
    names(maize),function(x){
                    paste("maize_de", i, ".csv", sep='')
    })
  out_filePath <- sapply(nam, function(x){
                     paste(outpath, x, sep='/')})
  write.csv(maize, file=out_filePath[i])
}

###LF: You can also use .txt instead of .csv 
###LF: Also, there are some errors in the naming of files, so I'm getting errors
```

#1 file with all SNPs with multiple positions in the genome (these need not be ordered in any particular way)

#For teosinte (Group = ZMPBA, ZMPIL, and ZMPJA in the third column of the fang_et_al_genotypes.txt file) we want 20 files in total:

#10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?
```{r}
for (i in 1:10) {
  teosinte <- filter(teosinte_fang_et_al_genotypes, Chromosome == i)
  teosinte <- rename(teosinte,NA,"?")
  teosinte <- arrange(teosinte, Position)
  outpath <- "/Users/yaweili/yawei-R-Assignment/teosinte_increasing_data/"
  nam <- sapply(
    names(teosinte),function(x){
                    paste("teosinte_in", i, ".csv", sep='')
    })
  out_filePath <- sapply(nam, function(x){
                     paste(outpath, x, sep='/')})
  write.csv(teosinte, file=out_filePath[i])
}
```


#10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -
```{r}
for (i in 1:10) {
  teosinte <- filter(teosinte_ded, Chromosome == i)
  teosinte <- arrange(teosinte, desc(Position))
  outpath <- "Users/yaweili/yawei-R-Assignment/teosinte_decreasing_data/"
  nam <- sapply(
    names(teosinte),function(x){
                    paste("teosinte_dd", i, ".csv", sep='')
    })
  out_filePath <- sapply(nam, function(x){
                     paste(outpath, x, sep='/')})
  write.csv(teosinte, file=out_filePath[i])

```

#In conclusion, there are 40 files produced.

###LF: For the loops, I couldn't generate the files because there were issues with the naming 

##Part II

#I need to use ggplot to visualize my data in this part. 
#But first I need to reshape the original data (make it tidy) using the melt command in the reshape2 package before attempting this part.
```{r}
if (!require("tidyverse")) install.packages("tidyverse")
library(tidyverse)
if (!require("reshape2")) install.packages("reshape2")
library(reshape2)
fang_et_al_genotypes <- read_tsv ("/Users/yaweili/yawei-R-Assignment/fang_et_al_genotypes.txt")
snp_position <- read_tsv ("/Users/yaweili/yawei-R-Assignment/snp_position.txt")
maize_fang_et_al_genotypes <- filter(fang_et_al_genotypes, Group == "ZMMLR" | Group == "ZMMIL" | Group == "ZMMMR")
teosinte_fang_et_al_genotypes <- filter(fang_et_al_genotypes, Group == "ZMPBA" | Group == "ZMPIL" | Group == "ZMPJA")
maize_fang_et_al_genotypes <- maize_fang_et_al_genotypes[,-c(1,2,3)]#cut head
maize_fang_et_al_genotypes <- t (maize_fang_et_al_genotypes)
teosinte_fang_et_al_genotypes <- teosinte_fang_et_al_genotypes[,-c(1,2,3)]#Guillotine
teosinte_fang_et_al_genotypes <- t (teosinte_fang_et_al_genotypes)
snp_position <- snp_position[,-c(2,5:15)]#Guillotine
arrange(snp_position, SNP_ID)
maize_fang_et_al_genotypes <- cbind(snp_position, maize_fang_et_al_genotypes)
teosinte_fang_et_al_genotypes <- cbind(snp_position, teosinte_fang_et_al_genotypes)
rm(fang_et_al_genotypes)#clear up
rm(snp_position)#clear up
setwd("/Users/yaweili/yawei-R-Assignment/")
```

#SNPs per chromosome

Plot the total number of SNPs in our dataset on each chromosome. Also plot the distribution of SNPs on chromosomes.

#Missing data and amount of heterozygosity

Create a new column to indicate whether a particular site is homozygous (has the same nucleotide on both chromosomes (i.e., A/A, C/C, G/G, T/T) or heterozygous (otherwise)). Make a graph that shows the proportion of homozygous and heterozygous sites as well as missing data in each sample (you won't be able to see the sample names). Make another graph that shows the same data for each group. Normalize the height of individual bars using one of the ggplot "position adjustments" options.

#Your own visualization

Visualize one other feature of the dataset. The choice is up to you!

###LF: I can see your graphs in your repository, and they're nicely generated, but I recommend including the script in your R markdown.